horovod/mxnet/mpi_ops.cc

```diff
@@ -20,60 +20,43 @@
 #include "../common/operations.h"
 #include "adapter.h"
 #include "cuda_util.h"
-#include "handle_manager.h"
 #include "mpi_ops.h"
 #include "ready_event.h"
 #include "tensor_util.h"
 
 namespace horovod {
 namespace mxnet {
 
-static HandleManager handle_manager;
-
 namespace {
-
 std::string GetOpName(std::string prefix, char* name) {
   if (name != nullptr) {
     return prefix + "." + std::string(name);
   }
   return std::string();
 }
-
-std::string GetOpNameHandle(std::string prefix, const std::string& name,
-                            int handle) {
-  if (name.length() != 0) {
-    return name;
-  }
-  return prefix + ".noname." + std::to_string(handle);
-}
-
 } // namespace
 
-int DoAllreduce(NDArray* tensor, NDArray* output, const std::string& name,
-                Callback cb) {
+void DoAllreduce(NDArray* tensor, NDArray* output, const std::string& name,
+                 Callback cb) {
   ThrowIfError(common::CheckInitialized());
 
-  auto handle = handle_manager.AllocateHandle(cb);
   auto device = TensorUtil::GetDevice(tensor);
   auto hvd_tensor = std::make_shared<MXTensor<NDArray>>(tensor);
   auto hvd_context = std::make_shared<MXOpContext<NDArray>>(device, output);
   auto hvd_output = std::make_shared<MXTensor<NDArray>>(output);
 
   auto enqueue_result =
       EnqueueTensorAllreduce(hvd_context, hvd_tensor, hvd_output, nullptr,
-                             GetOpNameHandle("allreduce", name, handle), device,
-                             [handle](const Status& status) {
-                               handle_manager.MarkDone(handle, status);
-                               handle_manager.ExecuteCallback(handle);
+                             name, device,
+                             [cb](const Status& status) {
+                               cb();
```

Comment:
can cb be NULL? We may need to check nullptr before calling it?

The mutex in the handle_manager was used to synchronize access to the map. Now we do not the lock here, right?

Reply:
cb here is not a pointer any more. I think it is created [here](https://github.com/apache/incubator-mxnet/blob/974a04cdb97fbabe079f79f1d9ebc2d14a793034/src/engine/threaded_engine.h#L349) in MXNet.
It is also not checked in parameter server design: https://github.com/apache/incubator-mxnet/blob/master/src/kvstore/kvstore_dist.h#L424

Reply:
My understanding of the mutex required is because we made a shared_ptr out of it in handle_manager: https://github.com/ctcyang/horovod/pull/15/files#diff-141f1184be39395e8f8708ff51a6fba7L27

Since we are not creating shared_ptr now, we no longer need to protect this resource, right?

Reply:
The mutex was to protect concurrent access to the unordered_map in handle_manager. shared_ptr was to track the ownership of the callback.

I think we should be fine.